File: src/main/java/com/murex/tbw/storage/JsonRepository.java
/*******************************************************************************
 *
 * Copyright (c) {2003-2019} Murex S.A.S. and its affiliates.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 *******************************************************************************/

package com.murex.tbw.storage;

import com.google.gson.stream.JsonReader;
import com.murex.tbw.domain.book.*;
import com.murex.tbw.domain.country.Country;
import com.murex.tbw.domain.country.Currency;
import com.murex.tbw.domain.country.Language;
import com.murex.tbw.purchase.Invoice;
import com.murex.tbw.purchase.PurchasedBook;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public final class JsonRepository implements Repository {
    private final File sourceFile;
    private final Map<Integer, Invoice> invoiceMap;

    public JsonRepository() {
        invoiceMap = new HashMap<>();
        sourceFile = new File(this.getClass().getClassLoader().getResource("repository.json").getFile());
        try {
            loadJsonData();
        } catch (FileNotFoundException e) {
            System.out.println(e.getMessage());
        }
    }

    private void loadJsonData() throws FileNotFoundException {
        JsonReader jsonReader = new JsonReader(new FileReader(sourceFile));
        try {
            jsonReader.beginArray();
            while (jsonReader.hasNext()) {
                Invoice invoice = null;
                String clientName = "";
                int id = 0;
                Country invoiceCountry;

                jsonReader.beginObject();
                while (jsonReader.hasNext()) {
                    String clientKeyName = jsonReader.nextName();
                    if (clientKeyName.equals("id")) {
                        id = jsonReader.nextInt();
                    } else if (clientKeyName.equals("client")) {
                        clientName = jsonReader.nextString();
                    } else if (clientKeyName.equals("country")) {
                        jsonReader.beginObject();
                        String countryName = "";
                        Language language = null;
                        Currency currency = null;
                        while (jsonReader.hasNext()) {
                            String countryKeyName = jsonReader.nextName();
                            if (countryKeyName.equals("name")) {
                                countryName = jsonReader.nextString();
                            } else if (countryKeyName.equals("language")) {
                                language = Language.valueOf(jsonReader.nextString());
                            } else if (countryKeyName.equals("currency")) {
                                currency = Currency.valueOf(jsonReader.nextString());
                            }
                        }
                        jsonReader.endObject();
                        invoiceCountry = new Country(countryName, currency, language);
                        invoice = new Invoice(id, clientName, invoiceCountry);
                    } else if (clientKeyName.equals("booksInBasket")) {
                        jsonReader.beginArray();
                        while (jsonReader.hasNext()) {
                            Book book;
                            String bookName = "";
                            double price = 0.0d;
                            Language bookLanguage = null;
                            Category bookCategory = null;
                            Author author = null;
                            int quantity = 0;
                            List<Genre> genres = new ArrayList<>();

                            jsonReader.beginObject();
                            while (jsonReader.hasNext()) {
                                String bookKeyName = jsonReader.nextName();
                                if (bookKeyName.equals("name")) {
                                    bookName = jsonReader.nextString();
                                } else if (bookKeyName.equals("price")) {
                                    price = jsonReader.nextDouble();
                                } else if (bookKeyName.equals("language")) {
                                    bookLanguage = Language.valueOf(jsonReader.nextString());
                                } else if (bookKeyName.equals("quantity")) {
                                    quantity = jsonReader.nextInt();
                                } else if (bookKeyName.equals("category")) {
                                    bookCategory = Category.valueOf(jsonReader.nextString());
                                } else if (bookKeyName.equals("author")) {
                                    jsonReader.beginObject();
                                    while (jsonReader.hasNext()) {
                                        String authorKeyName = jsonReader.nextName();
                                        String authorName = "";
                                        Country authorNationality = null;
                                        if (authorKeyName.equals("name")) {
                                            authorName = jsonReader.nextString();
                                        } else if (authorKeyName.equals("nationality")) {
                                            jsonReader.beginObject();
                                            String countryName = "";
                                            Language countryLanguage = null;
                                            Currency countryCurrency = null;
                                            while (jsonReader.hasNext()) {
                                                String authorNatKeyName = jsonReader.nextName();
                                                if (authorNatKeyName.equals("name")) {
                                                    countryName = jsonReader.nextString();
                                                } else if (authorNatKeyName.equals("language")) {
                                                    countryLanguage = Language.valueOf(jsonReader.nextString());
                                                } else if (authorNatKeyName.equals("currency")) {
                                                    countryCurrency = Currency.valueOf(jsonReader.nextString());
                                                }
                                            }
                                            authorNationality = new Country(countryName, countryCurrency, countryLanguage);
                                            jsonReader.endObject();
                                        }
                                        author = new Author(authorName, authorNationality);
                                    }
                                    jsonReader.endObject();
                                } else if (bookKeyName.equals("genre")) {
                                    jsonReader.beginArray();
                                    while (jsonReader.hasNext()) {
                                        genres.add(Genre.valueOf(jsonReader.nextString()));
                                    }
                                    jsonReader.endArray();
                                }
                            }
                            jsonReader.endObject();

                            if (bookCategory != null) {
                                book = new EducationalBook(bookName, price, author, bookLanguage, bookCategory);
                            } else {
                                book = new Novel(bookName, price, author, bookLanguage, genres);
                            }
                            invoice.addPurchasedBook(new PurchasedBook(book, quantity));
                        }
                        jsonReader.endArray();
                    }
                }
                addInvoice(invoice);
                jsonReader.endObject();
            }
            jsonReader.endArray();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void addInvoice(Invoice invoice) {
        invoiceMap.put(invoice.getId(), invoice);
    }

    @Override
    public Map<Integer, Invoice> getInvoiceMap() {
        return invoiceMap;
    }
}

File: typescript/src/com/murex/tbw/storage/JsonRepository.ts
import { Repository } from "./Repository";
import { Invoice } from "../purchase/Invoice";
import * as Immutable from "immutable";
import jsonData from "./repository.json";
import { Country } from "../domain/country/Country";
import { Currency } from "../domain/country/Currency";
import { Language } from "../domain/country/Language";
import { PurchasedBook } from "../purchase/PurchasedBook";
import { Novel } from "../domain/book/Novel";
import { EducationalBook } from "../domain/book/EducationalBook";
import { Author } from "../domain/book/Author";
import { Category } from "../domain/book/Category";
import { Genre } from "../domain/book/Genre";

export class JsonRepository implements Repository {
  private readonly invoices = new Map<number, Invoice>();

  public constructor() {
    this.loadJsonData();
  }

  addInvoice(invoice: Invoice): void {
    this.invoices.set(invoice.id, invoice);
  }

  getInvoiceMap(): Immutable.Map<number, Invoice> {
    return Immutable.Map<number, Invoice>(this.invoices);
  }

  private loadJsonData() {
    jsonData.forEach((jsonInvoice) => {
      const country = JsonRepository.toCountry(jsonInvoice.country);
      const invoice = new Invoice(jsonInvoice.client, country, jsonInvoice.id);

      jsonInvoice.booksInBasket.forEach((jsonPurchasedBook) => {
        const jsonAuthor = jsonPurchasedBook.author;
        const author = new Author(
          jsonAuthor.name,
          JsonRepository.toCountry(jsonAuthor.nationality)
        );

        let book;
        if (jsonPurchasedBook.genre !== undefined) {
          let genres = Immutable.Set<Genre>();
          jsonPurchasedBook.genre.forEach((jsonGenre) => {
            genres = genres.add(JsonRepository.toGenre(jsonGenre));
          });
          book = new Novel(
            jsonPurchasedBook.name,
            jsonPurchasedBook.price,
            author,
            JsonRepository.toLanguage(jsonPurchasedBook.language),
            genres
          );
        } else {
          book = new EducationalBook(
            jsonPurchasedBook.name,
            jsonPurchasedBook.price,
            author,
            JsonRepository.toLanguage(jsonPurchasedBook.language),
            JsonRepository.toCategory(jsonPurchasedBook.category)
          );
        }

        const purchasedBook = new PurchasedBook(
          book,
          jsonPurchasedBook.quantity
        );
        invoice.addPurchasedBook(purchasedBook);
      });

      this.addInvoice(invoice);
    });
  }

  private static toCountry({
    currency,
    language,
    name,
  }: {
    currency: string;
    language: string;
    name: string;
  }) {
    return new Country(
      name,
      JsonRepository.toCurrency(currency),
      JsonRepository.toLanguage(language)
    );
  }

  private static toLanguage(jsonLanguage: string) {
    switch (jsonLanguage) {
      case "ARABIC":
        return Language.Arabic;
      case "ENGLISH":
        return Language.English;
      case "FRENCH":
        return Language.French;
      case "GERMAN":
        return Language.German;
      case "JAPANESE":
        return Language.Japanese;
      case "MANDARIN":
        return Language.Mandarin;
      case "SPANISH":
        return Language.Spanish;
    }

    throw new Error("Unhandled language name: " + jsonLanguage);
  }

  private static toCurrency(jsonCurrency: string): Currency {
    switch (jsonCurrency) {
      case "US_DOLLAR":
        return Currency.Dollar;
      case "EURO":
        return Currency.Euro;
      case "RENMINBI":
        return Currency.Renminbi;
      case "POUND_STERLING":
        return Currency.PoundSterling;
      case "AUSTRALIAN_DOLLAR":
        return Currency.AustralianDollar;
      case "YEN":
        return Currency.Yen;
    }

    throw new Error("Unhandled currency name: " + jsonCurrency);
  }

  private static toCategory(jsonCategory: string) {
    switch (jsonCategory) {
      case "BUSINESS":
        return Category.BUSINESS;
      case "COMPUTER":
        return Category.COMPUTER;
      case "LANGUAGE":
        return Category.LANGUAGE;
    }

    throw new Error("Unhandled category name: " + jsonCategory);
  }

  private static toGenre(jsonGenre: string) {
    switch (jsonGenre) {
      case "HORROR_FICTION":
        return Genre.HORROR_FICTION;
      case "THRILLER":
        return Genre.THRILLER;
      case "DARK_FANTASY":
        return Genre.DARK_FANTASY;
      case "MYSTERY":
        return Genre.MYSTERY;
      case "ADVENTURE_FICTION":
        return Genre.ADVENTURE_FICTION;
      case "NON_FICTION":
        return Genre.NON_FICTION;
      case "ROMANCE":
        return Genre.ROMANCE;
    }

    throw new Error("Unhandled genre name: " + jsonGenre);
  }
}
